### PR TITLE
[Quant][fx][bc-breaking] Replace *custom_config_dict with config objects

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79066

Summary: Following https://github.com/pytorch/pytorch/pull/78452,
this commit replaces the following config dicts with python objects:

- prepare_custom_config_dict -> PrepareCustomConfig
- convert_custom_config_dict -> ConvertCustomConfig
- fuse_custom_config_dict -> FuseCustomConfig

This leads to better type safety and better user experience in
notebook settings due to improved auto completion.

BC-breaking notes:

Before:
```
from torch.ao.quantization.quantize_fx import prepare_fx, convert_fx

prepare_custom_config_dict = {
  "float_to_observed_custom_module_class": {
     "static": {
         FloatClass: ObservedClass
     }
  },
  "non_traceable_module_name": ["mod1", "mod2"],
  "non_traceable_module_class": [class1, class2],
  "input_quantized_idxs": [0, 1],
  "output_quantized_idxs": [0],
  "preserved_attributes": ["attr1", "attr2"],
}

convert_custom_config_dict = {
  "observed_to_quantized_custom_module_class": {
     "static": {
         FloatClass: ObservedClass
     }
  },
  "preserved_attributes": ["attr1", "attr2"],
}

model = prepare_fx(
    model,
    qconfig_mapping,
    example_inputs,
    prepare_custom_config_dict=prepare_custom_config_dict)

model(data)

model = convert_fx(model, convert_custom_config_dict=convert_custom_config_dict)
```

After:
```
from torch.ao.quantization.fx.custom_config import (
    PrepareCustomConfig,
    ConvertCustomConfig,
)
from torch.ao.quantization.quantize_fx import prepare_fx, convert_fx

prepare_custom_config = PrepareCustomConfig() \
    .set_float_to_observed_mapping(float_class, observed_class) \
    .set_non_traceable_module_names(["mod1", "mod2"]) \
    .set_non_traceable_module_classes([class1, class2]) \
    .set_input_quantized_indexes([0, 1]) \
    .set_output_quantized_indexes([0]) \
    .set_preserved_attributes(["attr1", "attr2"])

convert_custom_config = ConvertCustomConfig() \
    .set_observed_to_quantized_mapping(observed_class, quantized_class) \
    .set_preserved_attributes(["attr1", "attr2"])

model = prepare_fx(
    model,
    qconfig_mapping,
    example_inputs,
    prepare_custom_config=prepare_custom_config)

model(data)

model = convert_fx(model, convert_custom_config=convert_custom_config)
```

For backwards compatibility, prepare_fx, prepare_qat_fx, and
convert_fx will continue to accept Dicts, which will be converted
to the relevant *CustomConfig object internally.

Note that this commit does not modify existing tests to use the
new API; they will continue to pass in Dicts as before, which still
works but triggers a deprecation warning. This will be handled in
a future commit.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Differential Revision: [D37088095](https://our.internmc.facebook.com/intern/diff/D37088095)

cc @ezyang @SherlockNoMad @soumith

### CODE REVIEW

File: test/allowlist_for_publicAPI.json [hunk 1]
```diff
@@ -119,51 +119,6 @@
       "Tuple",
       "Union"
   ],
-  "torch.ao.quantization.fx.fuse": [
```
Comment (vkuzo) [inline test/allowlist_for_publicAPI.json:122]: just curious, what's the reason these changes are in this PR?

File: torch/ao/quantization/fx/custom_config.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    qconfig_mapping: Optional[QConfigMapping]
+    example_inputs: Tuple[Any, ...]
+    prepare_custom_config: Optional[PrepareCustomConfig]
+    # TODO: replace this with BackendConfig
+    backend_config_dict: Optional[Dict[str, Any]]
+
+
+class PrepareCustomConfig:
+    """
+    TODO: write this
```
Comment (vkuzo) [inline torch/ao/quantization/fx/custom_config.py:42]: before land? :)

Comment (andrewor14) [reply torch/ao/quantization/fx/custom_config.py:42]: Yes definitely :)

File: torch/ao/quantization/fx/custom_config.py [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+
+
+class PrepareCustomConfig:
+    """
+    Custom configuration for :func:`~torch.ao.quantization.quantize_fx.prepare_fx` and
+    :func:`~torch.ao.quantization.quantize_fx.prepare_qat_fx`.
+
+    The user can set custom configuration using the following methods:
+
+        `set_standalone_module_name`: sets the configuration for running a standalone module identified by name
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/custom_config.py:47]: running a standalone module --> preparing a standalone module for observation or quantization aware training?

File: torch/ao/quantization/fx/custom_config.py [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+        conf.set_preserved_attributes(fuse_custom_config_dict.get(PRESERVED_ATTRIBUTES_DICT_KEY, []))
+        return conf
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Convert this `FuseCustomConfig` to a dictionary with the items described in
+        :func:`~torch.ao.quantization.fx.custom_config.ConvertCustomConfig.from_dict`.
+        """
+        d: Dict[str, Any] = {}
+        for quant_type, observed_to_quantized_mapping in self.observed_to_quantized_mapping.items():
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/custom_config.py:410]: +1 for lint, probably need to remove

**Review by jerryzh168 — APPROVED** [review]
LGTM, this is a big change, thanks for the hard work! Please make sure the CI passes before landing, also will need to check fb-internal tests as well I think

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37088095).

File: test/allowlist_for_publicAPI.json (hunk 1 above)
Comment (anjali411) [reply test/allowlist_for_publicAPI.json:122]: nice cleanup!

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37088095).

**Pull request closed without merging**